integrate new ingredient parser with scraper

fernandalavalle opened the pull request:
Added a command line argument for parser URL and changed the ingredients module to make http post requests to this URL. 

Added ingredient tests for catching api errors and had to tweak test_html_parse.py to mock out the ingredient parser. 

A few formatting changes snuck in there when I formatted using yapf.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87)
<!-- Reviewable:end -->

mtlynch commented:
Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[scraper/ingredients.py, line 11 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU5x272p0EZboX_XFh:-LCU5x272p0EZboX_XFi:b-tusei3) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/scraper/ingredients.py#L11)):*
> ```Python
> 
> 
> def parse(ingredient_raw, url):
> ```

This is a case where it does make sense for us to convert to an object like

```python
class IngredientParser(object):

  def __init__(self, url):
    self._url = url

  def parse(ingredient_raw):
    ...
```

Otherwise the client has to keep track of the parse function, the URL, and the ingredient. If it's just an object, the client only has to worry about the URL and parser object.

---

*[scraper/ingredients.py, line 24 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU6Pn0-TThsx_qwmao:-LCU6Pn0-TThsx_qwmap:b-e9tbwj) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/scraper/ingredients.py#L24)):*
> ```Python
>     except urllib2.URLError as e:
>         raise IngredientParserApiError(
>             'Unable to reach ingredient parser: %s' % e)
> ```

`URLError` we probably don't want to wrap. We can't really add more context when we re-wrap it. If there's an exception, it's probably more obvious to clients what the problem is if they see an `HTTPError` than an `IngredientParserApiError`. We can just skip the `except URLError` and make it the caller's responsibility to deal with it.

Speaking of, can we add a docstring that enumerates the possible exceptions under a `Raises:` section?

---

*[scraper/ingredients.py, line 28 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU7PDyBIDTL1QkJqRb:-LCU7PDyBIDTL1QkJqRc:b-h10xhi) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/scraper/ingredients.py#L28)):*
> ```Python
>         raise IngredientParserApiError(
>             'Request to parse ingredient, %s, failed with: %s' %
>             (ingredient_raw, e))
> ```

If we get an `HTTPError`, we'll hopefully have a useful error message in the server's HTTP body (which we can access with `e.read()`). But if `e.read()` gives us back an empty string (which I think can happen if it's an HTTP 500), we should fall back to `e.message`.

---

*[scraper/ingredients.py, line 30 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCUB1JqCMrawaOlyEld:-LCUB1JqCMrawaOlyEle:b-tcq6ve) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/scraper/ingredients.py#L30)):*
> ```Python
>             (ingredient_raw, e))
> 
>     return json.loads(response.read())
> ```

I realized putting all the ingredient properties at the root level of the response kind of pigeonholes us if we want to return other metadata (e.g. which model version was used), so I changed it from:

```json
{  
   "comment":"young ones if possible",
   "name":"carrots",
   "quantity":1.0,
   "unit":"pound"
}
```

to:

```json
{  
   "ingredientParsed":{  
      "comment":"young ones if possible",
      "name":"carrots",
      "quantity":1.0,
      "unit":"pound"
   }
}
```

All we need to do here is change it to `json.loads(response.read())['ingredientParsed']`

---

*[tests/test_html_parse.py, line 77 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU8wZy36FN2PQiphNf:-LCU8wZy36FN2PQiphNg:b-rhaea5) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/tests/test_html_parse.py#L77)):*
> ```Python
>                         u'quantity': None,
>                         u'unit': None
>                     }],
> ```

Oh, you're not going to like this...

We actually don't need any field except for `name`.

`parse` should have the same interface as before where it receives a raw ingredient string and outputs a name string. Otherwise, we break downstream components that expect `ingredients` to be a list of strings. KetoHub just needs the names to make the recipes searchable and none of the other fields are relevant for searching (like I anticipate nobody wanting "tablespoons" or "finely diced" to match their search query).

---

*[tests/test_ingredients.py, line 36 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU7xUV5eOK5Ww9H3bw:-LCU7xUV5eOK5Ww9H3bx:bold4u7) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/tests/test_ingredients.py#L36)):*
> ```Python
>         self.mock_response.read.return_value = json.dumps({
>             "name":
>             "mock ingredient"
> ```

Double quotes -> single quotes

---

*[tests/test_ingredients.py, line 51 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU8pLb7kidkAYQQAAN:-LCU8pLb7kidkAYQQAAO:b-2qyz0l) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/tests/test_ingredients.py#L51)):*
> ```Python
>         self.mock_urlopen.side_effect = MockUrlError()
>         with self.assertRaises(ingredients.IngredientParserApiError):
>             ingredients.parse('mock ingredient', 'http://mockurl')
> ```

Can we add a success case as well?

---


*Comments from [Reviewable](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87)*
<!-- Sent from Reviewable.io -->

fernandalavalle commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[scraper/ingredients.py, line 11 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU5x272p0EZboX_XFh:-LCeVwnY6qZmSyEU4-xS:b-7rl0jd) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/scraper/ingredients.py#L11)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

This is a case where it does make sense for us to convert to an object like

```python
class IngredientParser(object):

  def __init__(self, url):
    self._url = url

  def parse(ingredient_raw):
    ...
```

Otherwise the client has to keep track of the parse function, the URL, and the ingredient. If it's just an object, the client only has to worry about the URL and parser object.
</blockquote></details>

done.

---

*[scraper/ingredients.py, line 24 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU6Pn0-TThsx_qwmao:-LCeWXx3AgDtClqAH84Y:b-pt3m1h) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/scraper/ingredients.py#L24)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

`URLError` we probably don't want to wrap. We can't really add more context when we re-wrap it. If there's an exception, it's probably more obvious to clients what the problem is if they see an `HTTPError` than an `IngredientParserApiError`. We can just skip the `except URLError` and make it the caller's responsibility to deal with it.

Speaking of, can we add a docstring that enumerates the possible exceptions under a `Raises:` section?
</blockquote></details>

I think wrapping it makes it very clear that it's a url error from trying to reach the parser versus any of the websites we're scraping. Why do you see the URLError differently from the HTTPError? 

added the docstring.

---

*[scraper/ingredients.py, line 28 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU7PDyBIDTL1QkJqRb:-LCegO6x5HHm3Ri4P0kE:b-rnrapz) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/scraper/ingredients.py#L28)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

If we get an `HTTPError`, we'll hopefully have a useful error message in the server's HTTP body (which we can access with `e.read()`). But if `e.read()` gives us back an empty string (which I think can happen if it's an HTTP 500), we should fall back to `e.message`.
</blockquote></details>

done. 

---

*[scraper/ingredients.py, line 30 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCUB1JqCMrawaOlyEld:-LCegMuw-9p7dqrfMzzx:b-7rl0jd) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/scraper/ingredients.py#L30)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

I realized putting all the ingredient properties at the root level of the response kind of pigeonholes us if we want to return other metadata (e.g. which model version was used), so I changed it from:

```json
{  
   "comment":"young ones if possible",
   "name":"carrots",
   "quantity":1.0,
   "unit":"pound"
}
```

to:

```json
{  
   "ingredientParsed":{  
      "comment":"young ones if possible",
      "name":"carrots",
      "quantity":1.0,
      "unit":"pound"
   }
}
```

All we need to do here is change it to `json.loads(response.read())['ingredientParsed']`
</blockquote></details>

done.

---

*[tests/test_html_parse.py, line 77 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU8wZy36FN2PQiphNf:-LCfl_So5F5ELqYZnnmG:b-lz2mji) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/tests/test_html_parse.py#L77)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

Oh, you're not going to like this...

We actually don't need any field except for `name`.

`parse` should have the same interface as before where it receives a raw ingredient string and outputs a name string. Otherwise, we break downstream components that expect `ingredients` to be a list of strings. KetoHub just needs the names to make the recipes searchable and none of the other fields are relevant for searching (like I anticipate nobody wanting "tablespoons" or "finely diced" to match their search query).
</blockquote></details>

not as bad as I was bracing for it to be!

---

*[tests/test_ingredients.py, line 36 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU7xUV5eOK5Ww9H3bw:-LCfl94G70DXhLaT8Vjl:b-7rl0jd) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/tests/test_ingredients.py#L36)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

Double quotes -> single quotes
</blockquote></details>

done.

---

*[tests/test_ingredients.py, line 51 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU8pLb7kidkAYQQAAN:-LCflK_4BvKtFk7ZErdL:b-4xiviu) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/tests/test_ingredients.py#L51)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

Can we add a success case as well?
</blockquote></details>

the first one is successful. what should I add?

---


*Comments from [Reviewable](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[scraper/ingredients.py, line 24 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU6Pn0-TThsx_qwmao:-LCiIYDLD1FPE2B7G1Bt:b-7xakep) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/scraper/ingredients.py#L24)):*
<details><summary><i>Previously, fernandalavalle wrote…</i></summary><blockquote>

I think wrapping it makes it very clear that it's a url error from trying to reach the parser versus any of the websites we're scraping. Why do you see the URLError differently from the HTTPError? 

added the docstring.
</blockquote></details>

I feel like it sort of mixes levels of abstraction. The parser is responsible for talking to the parse server and interpreting the server's answer. If the parse server successfully responds, but gives an application-layer error, it's the parser's job to interpret that error and handle it in a sensible way.

If it can't talk to the parse server because of something at the [transport layer](https://en.wikipedia.org/wiki/OSI_model) or below, it shouldn't try to handle that. In reality, there are tons of things that can go wrong during a network call, like an out of memory error, but the parser shouldn't try to rewrap every possible exception.

I agree that it does feel a little clearer seeing an error that specifically mentions the parser API, but I think realistically, you'd just see it in the stack trace anyway. Also the scraper only scrapes previously-scraped data stored on disk, so there shouldn't be any other URLs it's trying to contact.

---

*[scraper/ingredients.py, line 11 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiHoX62lKapfnuNL2C:-LCiHoX62lKapfnuNL2D:bkgndj2) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/scraper/ingredients.py#L11)):*
Something I've picked up from Go is avoiding "stutter naming":

>Avoid stutter. Since client code uses the package name as a prefix when referring to the package contents, the names for those contents need not repeat the package name. The HTTP server provided by the http package is called Server, not HTTPServer. Client code refers to this type as http.Server, so there is no ambiguity.

I think we can just call this `Parser` without any ambiguity.

---

*[scraper/ingredients.py, line 17 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiKc706wKm89CS19x2:-LCiKc706wKm89CS19x3:b-4hg6ic) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/scraper/ingredients.py#L17)):*
> ```Python
> 
>     def parse(self, ingredient_raw):
>         """Parses a raw text ingredient into an ingredient json object.
> ```

From the caller's perspective, it's just a dictionary. It doesn't have to know that it came from a JSON object.

---

*[scraper/ingredients.py, line 26 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiKvNw-kbfDsCE83Si:-LCiKvNw-kbfDsCE83Sj:b-ohv3ye) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/scraper/ingredients.py#L26)):*
> ```Python
>             A parsed ingredient, for example:
>                  {
>                    "ingredientParsed":{
> ```

As implemented, we're currently only returning the value of `ingredientParsed`.

But actually, why return anything more than the value of the name? That's all the offline scraper needs, so we might as well discard everything else as early as possible.

---

*[scraper/ingredients.py, line 38 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiP1ea6cdXFBERorJh:-LCiP1ea6cdXFBERorJi:b-69s23l) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/scraper/ingredients.py#L38)):*
> ```Python
>                 parser API.
>         """
> ```

extraneous newline

---

*[scraper/ingredients.py, line 54 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiLn4W7-sml6I-gKUD:-LCiLn4W7-sml6I-gKUE:b-ffg99l) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/scraper/ingredients.py#L54)):*
> ```Python
>             raise IngredientParserApiError(
>                 'Request to parse ingredient, %s, failed with: %s' %
>                 (ingredient_raw, e.message))
> ```

Oh I meant that we should try to read the contents of `e` because it has an HTTP body, so it should look like:

```python
error_body = e.read()
if not error_body:
  error_body = e.message
raise IngredientParserApiError(
  'Request to parse ingredient, %s, failed with: %s' %
  (ingredient_raw, error_body))
```

---

*[tests/test_html_parse.py, line 50 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiSPDbBnrjIhbjLkaV:-LCiSPDbBnrjIhbjLkaW:br8t5f2) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/tests/test_html_parse.py#L50)):*
> ```Python
>             u'quantity': None,
>             u'unit': None
>         }]
> ```

The particular return value from the ingredient parser doesn't actually matter because `html_parse` just passes it through without touching it. All we care about is that it comes back on the other side unchanged.

It generates a ton of tedious code to simulate the ingredient parser's behavior accurately for every recipe. It will be reduced once we change `parse` to only return the name field, but still.

I think we should just create a global fake `parse` function that essentially ignores the input and just returns `'fake parsed ingredient 1'`, `'parsed ingredient 2'` (where the count starts over from 1 in each test. It should be global to the test module so every test can just reuse the same dummy parser and not worry about setting up the mock for its particular test case.

We should keep asserting that `html_parse` called `parse` with the right ingredients from the HTML, but we can change the assertEqual to like:

```diff
'category': 'Dinner',
-'ingredients': [u'salt', u'water'],
+'ingredients': [u'fake parsed ingredient 1', u'fake parsed ingredient 2'],
'mainImage': 'http://a.b.com/img.jpg',
```

---

*[tests/test_html_parse.py, line 401 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiQOq588BshYbt0hC7:-LCiQOq588BshYbt0hC8:b-57a654) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/tests/test_html_parse.py#L401)):*
Suggest breaking them into one per line and adding `# yapf: disable`

Example: https://github.com/mtlynch/ingredient_parser/blob/49a10df824541a17778efa01e561a2983e81dd2c/tests/test_crfpp_translator.py#L16

Note that it's two spaces between the code and the `#` symbol:

>To improve legibility, these comments should be at least 2 spaces away from the code.

https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments

We normally wouldn't have to worry about whitespace, but since we're turning off YAPF for that line...

(ditto for every instance of multi-line ingredients)

---

*[tests/test_html_parse.py, line 418 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiTSRs3FvQ1VKQ1iM0:-LCiTSRs3FvQ1VKQ1iM1:b-185gw7) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/tests/test_html_parse.py#L418)):*
> ```Python
>                  u'4 oz cream cheese', u'2 tsp cumin', u'1 tsp oregano',
>                  u'1/4 tsp cayenne (optional)', u'Salt and Pepper to taste'
>              ])
> ```

Can we do this with `self.mock_ingredient_parser.assert_has_calls`? We also should avoid enforcing an order of calls because that's an implementation detail. `html_parse` should return the ingredients in order, but it doesn't necessarily have to parse them in order.

---

*[tests/test_ingredients.py, line 51 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU8pLb7kidkAYQQAAN:-LCiMrB25fw6DJc1nEMM:b-giwrul) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/tests/test_ingredients.py#L51)):*
<details><summary><i>Previously, fernandalavalle wrote…</i></summary><blockquote>

the first one is successful. what should I add?
</blockquote></details>

Oh right. I think I overlooked it because I didn't see it asserting on the return value, but it does that now.

---

*[tests/test_ingredients.py, line 9 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiNWRz1v_JuE40LGl8:-LCiNWRz1v_JuE40LGl9:bo02ozl) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/tests/test_ingredients.py#L9)):*
> ```Python
> 
> 
> class MockHttpError(urllib2.HTTPError):
> ```

I don't think we actually need to subclass it. I think we can do this when we want to simulate an error:

```python
mock_response = io.BytesIO('dummy server error')
urllib2.urlopen.side_effect = urllib2.HTTPError(
  url='mockurl', code=400, msg=None, hdrs={}, fp=mock_response)
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[scraper/ingredients.py, line 24 at r1](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCU6Pn0-TThsx_qwmao:-LCk1CYA-6jd6MDqRLqQ:b-wtu8w5) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5ad9f5e8909207681e952d9d7b3ee1095a58ac96/scraper/ingredients.py#L24)):*
<details><summary><i>Previously, fernandalavalle wrote…</i></summary><blockquote>

I was getting the URLError when I would forget to include the protocol for testing with the real ingredient parser server. Would a well formed URL not be in the application layer domain? It's a command-line argument and now an instance variable for the Parser. Should I just add validation in main? 
</blockquote></details>

Yeah we can do that, but let's do it in a separate PR.

---

*[scraper/ingredients.py, line 54 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiLn4W7-sml6I-gKUD:-LCk1NV8DWzpnnnvBBWe:b-o1sanq) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/scraper/ingredients.py#L54)):*
<details><summary><i>Previously, fernandalavalle wrote…</i></summary><blockquote>

Sorry, I thought you meant if there is no read() function then just use e.message. I get an attribute error when I call e.read(). 

More generally, why do you prefer read() so much more?
</blockquote></details>

Oh, they contain different different information. `e.read()` is the HTTP body that came back from the server.

So like here's what it looks like when there's an error that we defined in the server:

```python
 req = urllib2.Request(
  url='http://parser:8888/parse',
  data=json.dumps({'ingredient': 'A' * 5000}),
  headers={'Content-Type': 'application/json'}
  )
try:
  response = urllib2.urlopen(req)
except urllib2.HTTPError as e:
  print 'e.message=%s' % e.message
  print 'e.read()=%s' % e.read()
```

```
e.message=
e.read()="/parse request is too long to parse"
```

Oh, so I guess `e.message` is actually empty. I guess we should try `e.read()` and fall back to `str(e)`.

---

*[scraper/ingredients.py, line 25 at r3](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCjqT7I9jCpwS-QZYt0:-LCjqT7I9jCpwS-QZYt1:b2a36p0) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/2b52ed502023152eb5312cb18a0d07edf492908d/scraper/ingredients.py#L25)):*
> ```Python
>         Returns:
>             String of 'name' element of the ingredient, for example:
>                 'carrots's
> ```

Extra trailing s

---

*[tests/test_html_parse.py, line 51 at r3](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCk5oO06vK0pfTnVUIz:-LCk5oO06vK0pfTnVUJ-:b-vibpnt) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/2b52ed502023152eb5312cb18a0d07edf492908d/tests/test_html_parse.py#L51)):*
> ```Python
>                     'url': 'http://ignored.url',
>                     'category': 'Dinner',
>                     'ingredients': [  # yapf: disable
> ```

Oh, I'm surprised this works. I thought the yapf directive had to be at the end of the statement.

I'd rather move it to the the end (after the closing paren for `assertEqual`) because that's how I've been formatting it everywhere.

(ditto throughout)

---

*[tests/test_html_parse.py, line 260 at r3](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCk4TS76b4XUmj6KOvj:-LCk4TS76b4XUmj6KOvk:b-kpen1l) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/2b52ed502023152eb5312cb18a0d07edf492908d/tests/test_html_parse.py#L260)):*
> ```Python
> 
>         self.mock_ingredient_parser = mock.Mock(
>             side_effect=lambda x: mock_parse())
> ```

We can actually do this with a list comprehension that lets us delete `iterator` and `mock_parse`:

```python
self.mock_ingredient_parser = mock.Mock(
     side_effect=[u'fake parsed ingredient %d' % i for i in range(1, 100)])
```

---

*[tests/test_ingredients.py, line 31 at r3](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCk6wn-ETU6D58pmLIc:-LCk6wn-ETU6D58pmLId:b-oi61vi) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/2b52ed502023152eb5312cb18a0d07edf492908d/tests/test_ingredients.py#L31)):*
> ```Python
>         actual = self.parser.parse('mock ingredient')
>         self.assertTrue(isinstance(actual, dict))
>         self.assertDictEqual(actual, {'name': 'mock ingredient'})
> ```

I was really confused about how this was still working, but I realized `Parser.parse` never got updated so it's still returning the dictionary.

Can we update `parse` to return just the string and then update this test to match?

---


*Comments from [Reviewable](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87)*
<!-- Sent from Reviewable.io -->

fernandalavalle commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[scraper/ingredients.py, line 54 at r2](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCiLn4W7-sml6I-gKUD:-LClE3YYBzkYhcYddmuh:b-7rl0jd) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/5fb831fd847a090e08f6341731030eec885c3f38/scraper/ingredients.py#L54)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

Oh, they contain different different information. `e.read()` is the HTTP body that came back from the server.

So like here's what it looks like when there's an error that we defined in the server:

```python
 req = urllib2.Request(
  url='http://parser:8888/parse',
  data=json.dumps({'ingredient': 'A' * 5000}),
  headers={'Content-Type': 'application/json'}
  )
try:
  response = urllib2.urlopen(req)
except urllib2.HTTPError as e:
  print 'e.message=%s' % e.message
  print 'e.read()=%s' % e.read()
```

```
e.message=
e.read()="/parse request is too long to parse"
```

Oh, so I guess `e.message` is actually empty. I guess we should try `e.read()` and fall back to `str(e)`.


</blockquote></details>

done.

---

*[scraper/ingredients.py, line 25 at r3](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCjqT7I9jCpwS-QZYt0:-LClCyg2CJBVCauRTRy6:b-dmm5nv) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/2b52ed502023152eb5312cb18a0d07edf492908d/scraper/ingredients.py#L25)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

Extra trailing s
</blockquote></details>

whoops.

---

*[tests/test_html_parse.py, line 51 at r3](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCk5oO06vK0pfTnVUIz:-LClEkJn6YvWPbcRBqpq:b-7rl0jd) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/2b52ed502023152eb5312cb18a0d07edf492908d/tests/test_html_parse.py#L51)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

Oh, I'm surprised this works. I thought the yapf directive had to be at the end of the statement.

I'd rather move it to the the end (after the closing paren for `assertEqual`) because that's how I've been formatting it everywhere.

(ditto throughout)
</blockquote></details>

done.

---

*[tests/test_html_parse.py, line 260 at r3](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCk4TS76b4XUmj6KOvj:-LClCwnC-jB3qYg_Pnhp:b-7rl0jd) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/2b52ed502023152eb5312cb18a0d07edf492908d/tests/test_html_parse.py#L260)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

We can actually do this with a list comprehension that lets us delete `iterator` and `mock_parse`:

```python
self.mock_ingredient_parser = mock.Mock(
     side_effect=[u'fake parsed ingredient %d' % i for i in range(1, 100)])
```
</blockquote></details>

done.

---

*[tests/test_ingredients.py, line 31 at r3](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-LCk6wn-ETU6D58pmLIc:-LClFGj-9bxHopV88JYu:b-7rl0jd) ([raw file](https://github.com/mtlynch/ketohub_offline_scraper/blob/2b52ed502023152eb5312cb18a0d07edf492908d/tests/test_ingredients.py#L31)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

I was really confused about how this was still working, but I realized `Parser.parse` never got updated so it's still returning the dictionary.

Can we update `parse` to return just the string and then update this test to match?
</blockquote></details>

done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mtlynch/ketohub_offline_scraper/87#-:-LCnTFv-D1cJbiXBkbCh:bnfp4nl)*
<!-- Sent from Reviewable.io -->
